kernel/sched.c

```diff
@@ -798,17 +795,27 @@ void _impl_k_sleep(s32_t duration)
 	/* wait of 0 ms is treated as a 'yield' */
 	if (duration == 0) {
 		k_yield();
-		return;
+		return 0;
 	}
 
 	ticks = _TICK_ALIGN + _ms_to_ticks(duration);
+	expected_wakeup_time = ticks + z_tick_get_32();
+
+	compiler_barrier();
```

Comment:
Apparently compiler barrier is not needed here as `irq_lock` implies one. At least on ARM.

Reply:
But this is true only if there is no SMP...

Reply:
@pizi-nordic In such case I believe it must be fixed in `irq_lock`. Otherwise code will be broken in many places.

Reply:
Barrier removed. The irq_lock() either implies barrier (or ARM) or have to be fixed.